homeassistant/components/lutron/__init__.py

```diff
@@ -36,36 +37,74 @@
 )
 
 
-def setup(hass, base_config):
+async def async_setup(hass, base_config):
     """Set up the Lutron component."""
-    hass.data[LUTRON_BUTTONS] = []
-    hass.data[LUTRON_CONTROLLER] = None
-    hass.data[LUTRON_DEVICES] = {
-        "light": [],
-        "cover": [],
-        "switch": [],
-        "scene": [],
-        "binary_sensor": [],
-    }
-
+    hass.data.setdefault(DOMAIN, {})
     config = base_config.get(DOMAIN)
-    hass.data[LUTRON_CONTROLLER] = Lutron(
+
+    if config is not None:
+        hass.async_create_task(
+            hass.config_entries.flow.async_init(
+                DOMAIN, context={"source": config_entries.SOURCE_IMPORT}, data=config
+            )
+        )
+    return True
+
+
+class DummyArea:
+    """Fake area object.
+
+    This is used to tell all the lutron entities the names of the area they
+    belong to when we don't want to force the creation of those areas.
+    """
+
+    def __init__(self, name):
+        """Initialize the area."""
+        self.name = name
+        self.id = None
+
+
+async def async_setup_entry(hass, entry):
+    """Set up a single Lutron deployment."""
+    if not entry.data:
+        return False
+
+    config = entry.data
+    device_registry = await dr.async_get_registry(hass)
+    area_registry = await ar.async_get_registry(hass)
+    hass_data = {
+        LUTRON_BUTTONS: {},
+        LUTRON_CONTROLLER: None,
+        LUTRON_DEVICES: {
+            "light": [],
+            "cover": [],
+            "switch": [],
+            "scene": [],
+            "binary_sensor": [],
+        },
+    }
+    hass.data[DOMAIN][entry.entry_id] = hass_data
+    hass_data[LUTRON_CONTROLLER] = Lutron(
         config[CONF_HOST], config[CONF_USERNAME], config[CONF_PASSWORD]
     )
 
-    hass.data[LUTRON_CONTROLLER].load_xml_db()
-    hass.data[LUTRON_CONTROLLER].connect()
+    await hass.async_add_executor_job(hass_data[LUTRON_CONTROLLER].load_xml_db)
+    hass_data[LUTRON_CONTROLLER].connect()
     _LOGGER.info("Connected to main repeater at %s", config[CONF_HOST])
 
     # Sort our devices into types
-    for area in hass.data[LUTRON_CONTROLLER].areas:
+    for area in hass_data[LUTRON_CONTROLLER].areas:
+        if entry.options.get(CONF_ENABLE_AREAS, DEFAULT_ENABLE_AREAS):
+            hass_area = area_registry.async_get_or_create(area.name)
+        else:
+            hass_area = DummyArea(area.name)
```

Comment:
This seems weird. Creating areas and mixing them with dummy areas. I'm not sure we allow setting areas at all from integrations.